File: src/CSharp/Main/SpeedTest/SpeedTester.cs
using System;
using System.CodeDom.Compiler;
using System.IO;
using System.Linq;
using Benchmarking.Common;
using GCBurn.BurnTest;

namespace GCBurn.SpeedTest 
{
    public class SpeedTester
    {
        public const int PassCount = 30;
        public static TimeSpan DefaultDuration = TimeSpan.FromMilliseconds(1);
        public TimeSpan Duration = DefaultDuration;
        public IndentedTextWriter Writer = new IndentedTextWriter(Console.Out, "  ");
        
        public static SpeedTester New() => new SpeedTester();
        public static SpeedTester NewWarmup() => new SpeedTester() {
            Duration = TimeSpan.FromMilliseconds(5), 
            Writer = new IndentedTextWriter(TextWriter.Null),
        };
        
        public void Run()
        {
            var duration = Duration.TotalSeconds;
            var threadCount = ParallelRunner.ThreadCount;
            using (Writer.Section($"Test settings:")) {
                Writer.AppendMetric("Duration", duration * 1000, "ms");
                Writer.AppendMetric("Thread count", threadCount, "");
                Writer.AppendMetric("Unit size", UnitAllocator.UnitSize, "B");
            }
            Writer.AppendLine();

            var totalCount = 0L;
            for (var pass = 0; pass < PassCount; pass++) {
                var runner = ParallelRunner.New(i => new UnitAllocator(Duration));
                var allocators = runner.Run();
                totalCount = Math.Max(totalCount, allocators.Sum(a => a.AllocationCount));
            } 
            
            var totalSize = totalCount * UnitAllocator.UnitSize;
            var totalSizeWithOverhead = totalCount * (UnitAllocator.UnitSize + GarbageAllocator.ObjectSize);
            using (Writer.Section($"Allocation speed:")) {
                Writer.AppendMetric("Operations per second", totalCount / Sizes.Mega / duration, "M/s");
//                Writer.AppendMetric("Bytes per second", totalSize  / duration / Sizes.GB, "GB/s");
//                Writer.AppendMetric("Bytes per second (incl. overhead)", totalSizeWithOverhead  / duration / Sizes.GB, "GB/s");
            }
            Writer.AppendLine();
        }
    }
}

File: src/Go/speed_test/speed_tester.go
package speed_test

import (
	. "../common"
	"fmt"
	"time"
)

const PassCount = 30

var DefaultDuration = time.Duration(time.Millisecond)

type SpeedTester struct {
	NoOutput        bool
	Duration        time.Duration
	AllocationCount int64
}

func NewSpeedTester() *SpeedTester {
	t := &SpeedTester{
		NoOutput: false,
		Duration: DefaultDuration,
	}
	return t
}

func NewWarmupSpeedTester() *SpeedTester {
	t := NewSpeedTester()
	t.NoOutput = true
	t.Duration = time.Duration(5 * time.Millisecond)
	return t
}

func (t *SpeedTester) Run() {
	duration := t.Duration.Seconds()
	if !t.NoOutput {
		fmt.Printf("Test settings:\n")
		fmt.Printf("  Duration:     %v ms\n", int64(t.Duration.Seconds()/Milli))
		fmt.Printf("  Thread count: %v\n", ThreadCount)
		fmt.Printf("  Unit size:    %v B\n", UnitSize)
		fmt.Println()
	}

	totalCount := int64(0)
	for pass := 0; pass < PassCount; pass++ {
		runner := NewParallelRunner(func(i int) IActivity { return NewUnitAllocator(t.Duration) })
		activities := runner.Run()

		// Slice item casting
		allocators := make([]*UnitAllocator, 0)
		for _, a := range activities {
			allocators = append(allocators, a.(*UnitAllocator))
		}

		tc := int64(0)
		for _, a := range allocators {
			tc += a.AllocationCount
		}
		if totalCount < tc {
			totalCount = tc
		}
	}

	if t.NoOutput {
		return
	}

	fmt.Printf("Allocation speed:\n")
	fmt.Printf("  Operations per second: %.3f M/s\n", float64(totalCount)/duration/Mega)
	//fmt.Printf("  Bytes per second:      %.3f GB/s\n", float64(totalCount * UnitSize)/duration/GB)
	fmt.Println()
}
